src/backend/access/aocs/aocsam_handler.c

```diff
@@ -341,48 +352,41 @@ get_insert_descriptor(const Relation relation)
 
 	if (state->insertDesc == NULL)
 	{
-		List	*segments = NIL;
 		MemoryContext oldcxt;
 
+		/*
+		 * CBDB_PARALLEL:
+		 * Should not enable insertMultiFiles if the table is created by own transaction
+		 * or in utility mode.
+		 */
+		if (Gp_role != GP_ROLE_UTILITY &&
+			gp_appendonly_insert_files > 1 &&
+			!ShouldUseReservedSegno(relation, CHOOSE_MODE_WRITE))
+			state->insertMultiFiles = gp_appendonly_insert_files;
+
 		oldcxt = MemoryContextSwitchTo(aocoLocal.stateCxt);
 		state->insertDesc = aocs_insert_init(relation,
 									  ChooseSegnoForWrite(relation));
-		
-		dlist_init(&state->head);
-		dlist_head *head = &state->head;
-		dlist_push_tail(head, &state->insertDesc->node);
-
-		if (state->insertDesc->insertMultiFiles)
-		{
-			segments = lappend_int(segments, state->insertDesc->cur_segno);
-			for (int i = 0; i < gp_appendonly_insert_files - 1; i++)
-			{
-				next = aocs_insert_init(relation,
-												 ChooseSegnoForWriteMultiFile(relation, segments));
-				dlist_push_tail(head, &next->node);
-				segments = lappend_int(segments, next->cur_segno);
-			}
-			list_free(segments);
-		}
 
-		//* mark all insertDesc  placeholderInserted with false */
-        if (relationHasUniqueIndex(relation))
-        {
-            dlist_iter          iter;
-            dlist_foreach(iter, head)
-            {
-                AOCSInsertDesc insertDesc = (AOCSInsertDesc)dlist_container(AOCSInsertDescData, node, iter.cur);
-                insertDesc->placeholderInserted = false;
-            }
-        }
+		state->used_segment_files = list_make1_int(state->insertDesc->cur_segno);
+		dlist_init(&state->head);
+		dlist_push_tail(&state->head, &state->insertDesc->node);
 
 		MemoryContextSwitchTo(oldcxt);
 	}
 
 	/* switch insertDesc */
-	if (state->insertDesc->insertMultiFiles && state->insertDesc->range == gp_appendonly_insert_files_tuples_range)
+	if (state->insertMultiFiles && state->insertDesc->range == gp_appendonly_insert_files_tuples_range)
 	{
 		state->insertDesc->range = 0;
+
+		if (list_length(state->used_segment_files) < state->insertMultiFiles)
+		{
+			next = aocs_insert_init(relation, ChooseSegnoForWriteMultiFile(relation, state->used_segment_files));
+			dlist_push_tail(&state->head, &next->node);
+			state->used_segment_files = lappend_int(state->used_segment_files, next->cur_segno);
```

Comment:
can `next->cur_segno` already in  `state->used_segment_files` ?

Reply:
No, in principle. 
In what case it could be?

Reply:
I don't know, but this case used_segment_files will increase , exhaust insertMultiFiles.

Reply:
> used_segment_files will increase , exhaust insertMultiFiles.

Yeah, that's what we want to do, out goal is to ensure open a new variable segment file when needed, excluding  already opened ones, even by ourselves.